flang/lib/Optimizer/Transforms/RewriteLoop.cpp

```diff
@@ -44,7 +44,8 @@ class ScfLoopConv : public mlir::OpRewritePattern<fir::LoopOp> {
     auto step = loop.step();
     assert(low && high && step);
     // ForOp has different bounds semantics. Adjust upper bound.
-    auto adjustUp = rewriter.create<mlir::AddIOp>(loc, high, step);
+    auto one = rewriter.create<mlir::ConstantIndexOp>(loc, 1);
+    auto adjustUp = rewriter.create<mlir::AddIOp>(loc, high, one);
```

Comment:
It doesn't seem this works for countdown loops.

Can you comment on the failing case you are seeing?

Reply:
Oh my bad, I will try to fix that. Though I am not too sure how to verify that it is correct with countdown loops because they do not seem to be supported yet.

As for the failing case, when the stride is greater than one then, originally the upper bound was adjusted to be too high. For example in a do loop with a lower bound of 1, an upper bound of 6 and a stride of 2 the upper bound is adjusted to check that the IV is strictly lower than 8. So when the loop iterates it will loop with the IV having a value of 7 which is higher than the original upper bound.

So when running
```
      DO i = 1, 6, 2
         WRITE(*,*)  i
      END DO
```
f18 prints out
```
1
3
5
7
```
whereas gfortran prints out
```
1
3
5
```

Reply:
Thanks. I missed the link to the issue. My bad.

I think the upper bound (semi-open set) should be `high - low + step`. It looks like we lost the Sub op here.

Reply:
That makes sense and it seems to be working now with that change.